Item edit mode test

Joseph-Cashmore opened the pull request:
I have attempted to create an "edit mode" effect when the user clicks on an item. It only works to an extent.

The problem I ran into is as follows:

Attempt 1 - have some kind of toggle where the function is called onClick and onBlur; the user clicks on the item, it goes into edit mode (where they can change the item and delete it too) and then if they click off the item, the edit mode is removed.
The problem with this is that the app thinks that pressing the delete button is also a means of toggling the edit mode, so the item is toggled rather than being deleted.

Attempt 2 - Have the function called only onClick; the toggle function still works when clicking between items, but when clicking off it we get a weird behaviours like the delete button doesn't disappear. (This is the state the PR is currently in)

I must be going about this in the wrong way as it should be possible. Opening this up as a draft so anyone can have another go!

Commit:
font size adjustments

Commit:
padding adjustments

Commit:
updates to "completed" styling

I couldn't find a scenario with the use of a line separator that I liked, the heading is better.

Commit:
made new item form text more generic

Commit:
test fixes

Joseph-Cashmore committed:
Merge branch 'main' into ui-updates

Commit:
attempting an "isEditable" state to toggle different views

Getting some weird behaviours when item should lose focus.

Commit:
remove unnecessary style code

Commit:
test fix

Commit:
lint fixes

samstradling commented on src/components/ListItem/ListItem.js:
```diff
@@ -4,7 +4,7 @@ import { faTrash, faCheckCircle } from '@fortawesome/free-solid-svg-icons';
-const ListItem = ({ actions, keypressHandler, itemKey, text, isCompleted }) => {
+const ListItem = ({ actions, keypressHandler, itemKey, text, isCompleted, isEditable }) => {
   const [value, setValue] = useState(text);
   const mounted = useRef(false);
```
I think you'll want to keep the state about whether the item is being edited local to this component, so you can add to this component a new state item, something like this next to the existing `useState()`:
```js
const [isBeingEdited, setIsBeingEdited] = useState();
```

Then when we render the buttons we'll want to make that conditional on the item being edited:

```js
{isBeingEdited ? (
  <>
    <FontAwesomeIcon
      icon={faTrash}
      onClick={() => actions.deleteItemByKey(itemKey)}
      role="button"
      aria-label={`Delete item: ${text}`}
      data-testid={`delete-item-${text}`}
      tabIndex={0}
      className={styles.deleteIcon}
    />
    <FontAwesomeIcon
      icon={faCheckCircle}
      onClick={() => actions.toggleCompletionByKey(itemKey)}
      role="button"
      aria-label={`Added item: ${text}`}
      data-testid={`complete-item-${text}`}
      tabIndex={0}
      className={isCompleted ? (styles.completedIcon) : (styles.completeIcon)}
    />
  </>
) : null}
```

Then for the input box we can set the variable when it is focused or blured:
```
<input
  ... // existing fields
  onFocus={() => setIsBeingEdited(true)}
  onBlur={() => setIsBeingEdited(false)}
```

This will have the problem you mentioned the other day of being unable to click / tap the button because as soon as you do the input will blur and the button disappears, this can be fixed by adding a short delay:
```
onBlur={() => setTimeout(() => setIsBeingEdited(false), 20)}
```

This is a bit hacky, but should work ok

Joseph-Cashmore replied:
Ah ok wicked, thanks I'll give this a go! 

The delay is good to know, I was wanting something like that to make sure the edit mode change happened after everything else!

Commit:
new edit mode method (thanks sam!)

I found that the delay needed to be a certain length for this to work. All functions fine but not sure if there are ways of minimising the time

Commit:
fixed test

This feels like a bit of a cheat. The test isn't able to find the delete button in the current version, but if the element exists even with the class display none, it can be found

Joseph-Cashmore committed:
Merge branch 'main' into item-edit-mode-test
